20251205 support py312+

nucflash opened the pull request:
Hi, this is a follow up from PR #50; cc/  issue #49, #40. 

This PR provides:

* support for Python 3.10, 3.11, 3.12, 3.13, 3.14

* migrates from `setup.py` to `pyproject.toml` following the work by @fkiraly on XGBoostLSS

* adds `tox.ini` to support unit testing over multiple python versions using `tox`

* following XGBoostLSS, adds warnings to functions that require non-standard packages, e.g., `shap`, `optuna`. Skips tests that require these packages if the packages are not installed at test time. 

* adds a few more dirs to ignore in `.gitignore`

Important notes
1. *BREAKING CHANGE* `lightgbmlss.train()` no longer accepts the `feature_name` and `categorical_feature` parameters. This follows `lightgbm`'s decision to move these parameters to dataset creation. We need to define them in `lightgbm.Dataset`, and not in `lightgbm.train()`.

2.  `torch` should always be imported _before_ `lightgbm`. Failing to do so can lead to hard to debug segmentation faults. The reason lies in how the OpenMP is installed on a system. This is more likely to surface on macOS where OpenMP is installed using `brew`. `torch` brings its own version of OpenMP and uses that but lightgbm will try to find OpenMP in multiple locations and depending on the import sequence of torch and lightgbm, they can see different set of symbols leading to a segfault. I took the easier path and made sure that `torch` is imported before `lightgbm` everywhere in the code. You can read more about the issue and potential mitigations, [here](https://github.com/microsoft/LightGBM/issues/6595).

StatMixedML commented:
@nucflash Thanks for the PR!

I suggest we need a version range for lightgbm in


```python
dependencies = [
    "lightgbm",
    ...
    ]
```

nucflash commented on lightgbmlss/model.py:
```diff
@@ -1,3 +1,4 @@
+import torch
```
This is crucial to stay at the top. It solves seg fault on macOS due to OpenMP libs (see main text of PR).

nucflash commented:
Hi @StatMixedML , thanks for the comments. Nice findings from Claude. I made some changes:

* added version constraints to lightgbm package 4<=lgb<5
* fixed typos and cruft from aligning with xgboostlss (as pointed out from Claude)
* replaced setup.py with pyproject.toml (setup.py is deleted)

Let me know what you think. Thanks again!
